tests/model_fields/test_jsonfield.py

```diff
@@ -0,0 +1,131 @@
+import uuid
+
+from django.db import connection, models, transaction
+from django.db.utils import IntegrityError
+from django.test import TestCase
+
+from .models import JSONCustomEncoderDecoderModel, JSONModel
+
+
+class JSONFieldTests(TestCase):
+
+    def _test_scalar_value(self, value):
+        obj = JSONModel(value=value)
+        # Oracle Database doesn't allow scalar values
+        if connection.vendor == 'oracle':
+            with transaction.atomic():
+                self.assertRaises(IntegrityError, obj.save)
+        else:
+            obj.save()
+            obj = JSONModel.objects.get(id=obj.id)
+            self.assertEqual(obj.value, value)
+
+    def test_bool_value(self):
+        self._test_scalar_value(True)
+        self._test_scalar_value(False)
+
+    def test_int_value(self):
+        self._test_scalar_value(123456)
+
+    def test_float_value(self):
+        self._test_scalar_value(1234.56)
+
+    def test_str_value(self):
+        self._test_scalar_value('A string')
+
+    def test_dict_value(self):
+        value = {'name': 'John', 'age': 20, 'height': 180.3}
+        obj = JSONModel.objects.create(value=value)
+        obj = JSONModel.objects.get(id=obj.id)
+        self.assertEqual(obj.value, value)
+
+    def test_list_value(self):
+        value = ['John', 20, 180.3]
+        obj = JSONModel.objects.create(value=value)
+        obj = JSONModel.objects.get(id=obj.id)
+        self.assertEqual(obj.value, value)
+
+    def test_nested_value(self):
+        value = {
+            'name': 'John',
+            'age': 20,
+            'pets': [
+                {'name': 'Kit', 'type': 'cat', 'age': 2},
+                {'name': 'Max', 'type': 'dog', 'age': 1}
+            ],
+            'courses': [
+                ['A1', 'A2', 'A3'], ['B1', 'B2'], ['C1']
+            ]
+        }
+        obj = JSONModel.objects.create(value=value)
+        obj = JSONModel.objects.get(id=obj.id)
+        self.assertEqual(obj.value, value)
+
+    def _test_mutable_default_value(self, model):
+        errors = model.check(actually_check=True)
+        self.assertEqual(len(errors), 1)
+        self.assertIn('default should not be a mutable object', errors[0].msg)
+
+    def test_dict_default_value(self):
+        class JSONDictDefaultModel(models.Model):
+            value = models.JSONField(default={'foo': 'bar'})
+
+        self._test_mutable_default_value(JSONDictDefaultModel)
+
+    def test_list_default_value(self):
+        class JSONListDefaultModel(models.Model):
+            value = models.JSONField(default=['foo', 'bar', 123])
+
+        self._test_mutable_default_value(JSONListDefaultModel)
+
+    def _test_valid_default_value(self, model):
+        errors = model.check(actually_check=True)
+        self.assertEqual(len(errors), 0)
```

Comment:
Should we move this code to `_test_valid_default_values(self)` as just one test with `subTest()` ? we would define all test cases with tuple. (same `_test_mutable_default_value`)

Reply:
Good idea. I've also done so with `test_scalar_values`.

Reply:
I would use just one `Model` for `test_valid_default_value ` and change its value field's default value under subtest.

Reply:
you can extend `valid_models` items as tuple. 
```
valid_models = [
            (JSONCallableDefault, {'it': 'works'})
            ...
```

Reply:
Yeah, I tried doing that before, but apparently creating a new model with the same name causes issues with model registration. I'm not sure how to avoid that. Any clues?

Reply:
```
default_values =(
    {'it': 'works'},
    True,
    123
)
class JSONModel(models.Model):
    value = models.JSONField()
for default_value in default_values:
    with self.subTest(default_value=default_value):
        field = JSONModel._meta.get_field('value')
        field.default = default_value
        errors = field.check()
        self.assertEqual(len(errors), 0)
```

Reply:
I've noticed that `valid_models` is not necessary for this case.

Reply:
Thanks for the tip!